src/proto/codec/mod.rs

```diff
@@ -433,6 +447,35 @@ impl PacketCodecInner {
         }
     }
 
+    fn is_at_packet_boundary(&self) -> bool {
+        match self {
+            PacketCodecInner::Plain(c) => c.chunk_decoder == ChunkDecoder::Idle,
+            PacketCodecInner::Comp(c) => {
+                c.in_buf.is_empty()
+                    && c.out_buf.is_empty()
+                    && c.comp_decoder == CompDecoder::Idle
+                    && c.plain_codec.chunk_decoder == ChunkDecoder::Idle
+            }
+        }
+    }
+
+    fn save_state(&self) -> PacketCodecState {
```

Comment:
I went back and forth on this; I wanted to emphasize that it was more like a snapshot that you could roll back to, rather than something you could construct and pass in. If you want, I can change it to `get`/`set`, I don't feel strongly about it.

Reply:
I like `snapshot_state`. Was thinking of naming it `restore_state` and `get_state` but those don't seem like they belong together. So `snapshot_state` and `restore_state` sounds best to me